Regression of Compose Searchbar (#237)

```
java.lang.RuntimeException: Unable to destroy activity {moe.tarsin.ehviewer.debug/com.hippo.ehviewer.ui.MainActivity}: java.lang.NullPointerException
	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:5613)
	at android.app.ActivityThread.handleDestroyActivity(ActivityThread.java:5645)
	at android.app.ActivityThread.handleRelaunchActivityInner(ActivityThread.java:5923)
	at android.app.ActivityThread.handleRelaunchActivity(ActivityThread.java:5842)
	at android.app.servertransaction.ActivityRelaunchItem.execute(ActivityRelaunchItem.java:76)
	at android.app.servertransaction.ActivityTransactionItem.execute(ActivityTransactionItem.java:45)
	at android.app.servertransaction.TransactionExecutor.executeCallbacks(TransactionExecutor.java:139)
	at android.app.servertransaction.TransactionExecutor.execute(TransactionExecutor.java:96)
	at android.app.ActivityThread$H.handleMessage(ActivityThread.java:2443)
	at android.os.Handler.dispatchMessage(Handler.java:106)
	at android.os.Looper.loopOnce(Looper.java:205)
	at android.os.Looper.loop(Looper.java:294)
	at android.app.ActivityThread.main(ActivityThread.java:8177)
	at java.lang.reflect.Method.invoke(Native Method)
	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:552)
	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:971)
Caused by: java.lang.NullPointerException
	at com.hippo.ehviewer.ui.scene.GalleryListScene.getBinding(GalleryListScene.kt:190)
	at com.hippo.ehviewer.ui.scene.GalleryListScene.onDestroyView(GalleryListScene.kt:451)
	at androidx.fragment.app.Fragment.performDestroyView(Fragment.java:3362)
	at androidx.fragment.app.FragmentStateManager.destroyFragmentView(FragmentStateManager.java:788)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:344)
	at androidx.fragment.app.SpecialEffectsController$FragmentStateManagerOperation.complete$fragment_release(SpecialEffectsController.kt:832)
	at androidx.fragment.app.SpecialEffectsController$Operation.cancel(SpecialEffectsController.kt:656)
	at androidx.fragment.app.SpecialEffectsController.forceCompleteAllOperations(SpecialEffectsController.kt:356)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3188)
	at androidx.fragment.app.FragmentManager.dispatchDestroyView(FragmentManager.java:3131)
	at androidx.fragment.app.Fragment.performDestroyView(Fragment.java:3355)
	at androidx.fragment.app.FragmentStateManager.destroyFragmentView(FragmentStateManager.java:788)
	at androidx.fragment.app.FragmentStateManager.moveToExpectedState(FragmentStateManager.java:344)
	at androidx.fragment.app.SpecialEffectsController$FragmentStateManagerOperation.complete$fragment_release(SpecialEffectsController.kt:832)
	at androidx.fragment.app.SpecialEffectsController$Operation.cancel(SpecialEffectsController.kt:656)
	at androidx.fragment.app.SpecialEffectsController.forceCompleteAllOperations(SpecialEffectsController.kt:356)
	at androidx.fragment.app.FragmentManager.dispatchStateChange(FragmentManager.java:3188)
	at androidx.fragment.app.FragmentManager.dispatchDestroy(FragmentManager.java:3139)
	at androidx.fragment.app.FragmentController.dispatchDestroy(FragmentController.java:346)
	at androidx.fragment.app.FragmentActivity.onDestroy(FragmentActivity.java:258)
	at androidx.appcompat.app.AppCompatActivity.onDestroy(AppCompatActivity.java:283)
	at android.app.Activity.performDestroy(Activity.java:8876)
	at android.app.Instrumentation.callActivityOnDestroy(Instrumentation.java:1491)
	at android.app.ActivityThread.performDestroyActivity(ActivityThread.java:5600)
	... 15 more
```

--- a/app/src/main/kotlin/com/hippo/ehviewer/ui/MainActivity.kt
+++ b/app/src/main/kotlin/com/hippo/ehviewer/ui/MainActivity.kt
@@ -56,13 +56,10 @@ import androidx.compose.material3.ModalNavigationDrawer
 import androidx.compose.material3.NavigationDrawerItem
 import androidx.compose.material3.Text
 import androidx.compose.material3.rememberDrawerState
-import androidx.compose.runtime.CompositionContext
-import androidx.compose.runtime.DisposableEffect
 import androidx.compose.runtime.LaunchedEffect
 import androidx.compose.runtime.currentRecomposeScope
 import androidx.compose.runtime.getValue
 import androidx.compose.runtime.mutableStateOf
-import androidx.compose.runtime.rememberCompositionContext
 import androidx.compose.runtime.rememberCoroutineScope
 import androidx.compose.runtime.setValue
 import androidx.compose.ui.Modifier
@@ -228,8 +225,6 @@ class MainActivity : EhActivity() {
         recomposeFlow.tryEmit(Unit)
     }
 
-    var compositionContext: CompositionContext? = null
-
     override fun onCreate(savedInstanceState: Bundle?) {
         installSplashScreen()
         super.onCreate(savedInstanceState)
@@ -242,13 +237,6 @@ class MainActivity : EhActivity() {
             fun isSelected(id: Int) = ::navController.isInitialized && id == navController.currentDestination?.id
             fun closeDrawer() = scope.launch { drawerState.close() }
 
-            val currentCompositionContext = rememberCompositionContext()
-            DisposableEffect(Unit) {
-                compositionContext = currentCompositionContext
-                onDispose {
-                    compositionContext = null
-                }
-            }
             LaunchedEffect(Unit) {
                 runSuspendCatching {
                     withIOContext {

--- a/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/BaseScene.kt
+++ b/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/BaseScene.kt
@@ -17,28 +17,14 @@ package com.hippo.ehviewer.ui.scene
 
 import android.content.res.Resources.Theme
 import android.os.Bundle
-import android.os.Parcelable
-import android.util.SparseArray
-import android.view.LayoutInflater
 import android.view.View
 import androidx.annotation.StringRes
 import androidx.core.view.SoftwareKeyboardControllerCompat
 import androidx.fragment.app.Fragment
-import androidx.lifecycle.ViewModelStoreOwner
-import androidx.lifecycle.setViewTreeLifecycleOwner
-import androidx.lifecycle.setViewTreeViewModelStoreOwner
-import androidx.savedstate.SavedStateRegistryOwner
-import androidx.savedstate.setViewTreeSavedStateRegistryOwner
-import com.google.android.material.sidesheet.SideSheetDialog
 import com.hippo.ehviewer.ui.MainActivity
-import com.hippo.ehviewer.util.getSparseParcelableArrayCompat
 import rikka.core.res.resolveDrawable
 
 abstract class BaseScene : Fragment() {
-    private var drawerView: View? = null
-    private var drawerViewState: SparseArray<Parcelable>? = null
-    private var sideSheetDialog: SideSheetDialog? = null
-
     val isDrawerLocked
         get() = mainActivity?.drawerLocked == true
 
@@ -52,9 +38,6 @@ abstract class BaseScene : Fragment() {
 
     fun openDrawer() = mainActivity?.openDrawer()
 
-    fun openSideSheet() = sideSheetDialog!!.show()
-    fun closeSideSheet() = sideSheetDialog!!.dismiss()
-
     fun showTip(message: CharSequence?, length: Int) {
         mainActivity?.showTip(message!!, length)
     }
@@ -64,35 +47,6 @@ abstract class BaseScene : Fragment() {
     }
 
     open val enableDrawerGestures = false
-
-    private fun createDrawerView(savedInstanceState: Bundle?): View? {
-        drawerView = onCreateDrawerView(layoutInflater)
-        if (drawerView != null) {
-            val saved = drawerViewState ?: savedInstanceState?.getSparseParcelableArrayCompat(KEY_DRAWER_VIEW_STATE)
-            saved?.let {
-                drawerView!!.restoreHierarchyState(saved)
-            }
-        }
-        return drawerView
-    }
-
-    open fun onCreateDrawerView(inflater: LayoutInflater): View? {
-        return null
-    }
-
-    private fun destroyDrawerView() {
-        if (drawerView != null) {
-            drawerViewState = SparseArray()
-            drawerView!!.saveHierarchyState(drawerViewState)
-        }
-        onDestroyDrawerView()
-        drawerView = null
-    }
-
-    open fun onDestroyDrawerView() {
-        sideSheetDialog?.dismiss()
-        sideSheetDialog = null
-    }
 
     override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
         super.onViewCreated(view, savedInstanceState)
@@ -105,38 +59,15 @@ abstract class BaseScene : Fragment() {
             lockDrawer()
         }
         hideSoftInput()
-        createDrawerView(savedInstanceState)?.let {
-            sideSheetDialog = SideSheetDialog(requireContext()).apply {
-                window?.decorView?.apply {
-                    val owner = viewLifecycleOwner
-                    setViewTreeLifecycleOwner(owner)
-                    setViewTreeViewModelStoreOwner(owner as ViewModelStoreOwner)
-                    setViewTreeSavedStateRegistryOwner(owner as SavedStateRegistryOwner)
-                }
-                setContentView(it)
-            }
-        }
         mainActivity?.recompose()
-    }
-
-    override fun onDestroyView() {
-        super.onDestroyView()
-        destroyDrawerView()
     }
 
     val mainActivity: MainActivity?
         get() = activity as? MainActivity
 
-    fun hideSoftInput() = activity?.window?.decorView?.run { SoftwareKeyboardControllerCompat(this) }?.hide()
-
-    override fun onSaveInstanceState(outState: Bundle) {
-        super.onSaveInstanceState(outState)
-        if (drawerView != null) {
-            drawerViewState = SparseArray()
-            drawerView!!.saveHierarchyState(drawerViewState)
-            outState.putSparseParcelableArray(KEY_DRAWER_VIEW_STATE, drawerViewState)
-        }
-    }
+    private fun hideSoftInput() = activity?.window?.decorView?.run {
+        SoftwareKeyboardControllerCompat(this)
+    }?.hide()
 
     val theme: Theme
         get() = requireActivity().theme

--- a/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/DownloadsScene.kt
+++ b/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/DownloadsScene.kt
@@ -370,8 +370,8 @@ class DownloadsScene :
         super.onSearchViewExpanded()
     }
 
-    override fun onDestroyView() {
-        super.onDestroyView()
+    override fun onRelease() {
+        super.onRelease()
         binding.recyclerView.stopScroll()
         mViewTransition = null
         mAdapter = null

--- a/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/FavoritesScene.kt
+++ b/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/FavoritesScene.kt
@@ -73,7 +73,6 @@ import com.hippo.ehviewer.client.data.FavListUrlBuilder
 import com.hippo.ehviewer.client.ehUrl
 import com.hippo.ehviewer.databinding.SceneFavoritesBinding
 import com.hippo.ehviewer.ui.CommonOperations
-import com.hippo.ehviewer.ui.MainActivity
 import com.hippo.ehviewer.ui.legacy.AddDeleteDrawable
 import com.hippo.ehviewer.ui.legacy.BaseDialogBuilder
 import com.hippo.ehviewer.ui.legacy.FabLayout
@@ -429,8 +428,8 @@ class FavoritesScene : SearchBarScene() {
         )
     }
 
-    override fun onDestroyView() {
-        super.onDestroyView()
+    override fun onRelease() {
+        super.onRelease()
         binding.recyclerView.stopScroll()
         mAdapter = null
         _binding = null
@@ -467,9 +466,6 @@ class FavoritesScene : SearchBarScene() {
                 }
             }
         }
-    }.apply {
-        val compositionContext = (requireActivity() as MainActivity).compositionContext!!
-        setParentCompositionContext(compositionContext)
     }
 
     override fun onSearchViewExpanded() {

--- a/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/GalleryListScene.kt
+++ b/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/GalleryListScene.kt
@@ -445,8 +445,8 @@ class GalleryListScene : SearchBarScene() {
         return binding
     }
 
-    override fun onDestroyView() {
-        super.onDestroyView()
+    override fun onRelease() {
+        super.onRelease()
         stateBackPressedCallback.remove()
         binding.recyclerView.stopScroll()
         _binding = null

--- a/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/SearchBarScene.kt
+++ b/app/src/main/kotlin/com/hippo/ehviewer/ui/scene/SearchBarScene.kt
@@ -1,9 +1,12 @@
 package com.hippo.ehviewer.ui.scene
 
 import android.os.Bundle
+import android.os.Parcelable
+import android.util.SparseArray
 import android.view.LayoutInflater
 import android.view.View
 import android.view.ViewGroup
+import androidx.annotation.CallSuper
 import androidx.compose.animation.AnimatedVisibility
 import androidx.compose.animation.scaleIn
 import androidx.compose.animation.scaleOut
@@ -44,21 +47,29 @@ import androidx.compose.runtime.getValue
 import androidx.compose.runtime.mutableIntStateOf
 import androidx.compose.runtime.mutableStateOf
 import androidx.compose.runtime.remember
+import androidx.compose.runtime.rememberCompositionContext
 import androidx.compose.runtime.rememberCoroutineScope
 import androidx.compose.runtime.setValue
 import androidx.compose.ui.Alignment
 import androidx.compose.ui.Modifier
 import androidx.compose.ui.graphics.Color
 import androidx.compose.ui.layout.layout
+import androidx.compose.ui.platform.ComposeView
 import androidx.compose.ui.platform.LocalContext
 import androidx.compose.ui.platform.LocalDensity
 import androidx.compose.ui.res.stringResource
 import androidx.compose.ui.unit.dp
 import androidx.compose.ui.viewinterop.AndroidViewBinding
 import androidx.core.view.updatePadding
+import androidx.lifecycle.ViewModelStoreOwner
+import androidx.lifecycle.setViewTreeLifecycleOwner
+import androidx.lifecycle.setViewTreeViewModelStoreOwner
 import androidx.recyclerview.widget.RecyclerView
 import androidx.recyclerview.widget.RecyclerView.OnScrollListener
+import androidx.savedstate.SavedStateRegistryOwner
+import androidx.savedstate.setViewTreeSavedStateRegistryOwner
 import androidx.viewbinding.ViewBinding
+import com.google.android.material.sidesheet.SideSheetDialog
 import com.hippo.ehviewer.EhApplication.Companion.searchDatabase
 import com.hippo.ehviewer.R
 import com.hippo.ehviewer.Settings
@@ -68,6 +79,7 @@ import com.hippo.ehviewer.dao.SearchDao
 import com.hippo.ehviewer.ui.MainActivity
 import com.hippo.ehviewer.ui.tools.LocalDialogState
 import com.hippo.ehviewer.util.findActivity
+import com.hippo.ehviewer.util.getSparseParcelableArrayCompat
 import com.jamal.composeprefs3.ui.ifNotNullThen
 import com.jamal.composeprefs3.ui.ifTrueThen
 import eu.kanade.tachiyomi.util.lang.launchIO
@@ -315,6 +327,10 @@ fun SearchBarScreen(
 }
 
 abstract class SearchBarScene : BaseScene() {
+    private var drawerView: View? = null
+    private var drawerViewState: SparseArray<Parcelable>? = null
+    private var sideSheetDialog: SideSheetDialog? = null
+
     private var mSuggestionProvider: SuggestionProvider? = null
     private var onApplySearch: (String) -> Unit = {}
     var initialQuery = ""
@@ -336,6 +352,7 @@ abstract class SearchBarScene : BaseScene() {
             }
         }
         return ComposeWithMD3 {
+            val compositionContext = rememberCompositionContext()
             val density = LocalDensity.current
             val fabPadding = with(density) { 16.dp.roundToPx() }
             val margin = with(density) { 8.dp.roundToPx() }
@@ -357,9 +374,24 @@ abstract class SearchBarScene : BaseScene() {
                 AndroidViewBinding(
                     modifier = Modifier.fillMaxSize(),
                     factory = { inflater, parent, _ ->
-                        onCreateViewWithToolbar(inflater, parent, savedInstanceState).apply {
+                        onCreateViewWithToolbar(inflater, parent, savedInstanceState).also {
                             recyclerView.addOnScrollListener(onScrollListener)
+                            createDrawerView(savedInstanceState)?.apply {
+                                if (this is ComposeView) {
+                                    val owner = viewLifecycleOwner
+                                    setViewTreeLifecycleOwner(owner)
+                                    setViewTreeViewModelStoreOwner(owner as ViewModelStoreOwner)
+                                    setViewTreeSavedStateRegistryOwner(owner as SavedStateRegistryOwner)
+                                    setParentCompositionContext(compositionContext)
+                                }
+                                sideSheetDialog = SideSheetDialog(parent.context).also {
+                                    it.setContentView(this)
+                                }
+                            }
                         }
+                    },
+                    onRelease = {
+                        onRelease()
                     },
                 ) {
                     fabLayout.updatePadding(bottom = fabPadding + bottomPadding)
@@ -375,6 +407,49 @@ abstract class SearchBarScene : BaseScene() {
         container: ViewGroup,
         savedInstanceState: Bundle?,
     ): ViewBinding
+
+    fun openSideSheet() = sideSheetDialog!!.show()
+    fun closeSideSheet() = sideSheetDialog!!.dismiss()
+
+    private fun createDrawerView(savedInstanceState: Bundle?): View? {
+        drawerView = onCreateDrawerView(layoutInflater)?.apply {
+            val saved = drawerViewState ?: savedInstanceState?.getSparseParcelableArrayCompat(KEY_DRAWER_VIEW_STATE)
+            saved?.let {
+                restoreHierarchyState(it)
+            }
+        }
+        return drawerView
+    }
+
+    open fun onCreateDrawerView(inflater: LayoutInflater): View? = null
+
+    private fun destroyDrawerView() {
+        drawerView?.let {
+            drawerViewState = SparseArray()
+            it.saveHierarchyState(drawerViewState)
+        }
+        onDestroyDrawerView()
+        drawerView = null
+    }
+
+    open fun onDestroyDrawerView() {
+        sideSheetDialog?.dismiss()
+        sideSheetDialog = null
+    }
+
+    override fun onSaveInstanceState(outState: Bundle) {
+        super.onSaveInstanceState(outState)
+        drawerView?.let {
+            drawerViewState = SparseArray()
+            it.saveHierarchyState(drawerViewState)
+            outState.putSparseParcelableArray(KEY_DRAWER_VIEW_STATE, drawerViewState)
+        }
+    }
+
+    @CallSuper
+    open fun onRelease() {
+        destroyDrawerView()
+    }
 
     @Composable
     abstract fun TrailingIcon()

